### DISCUSSION TITLE
2.0.1 purge problem(?)

### DISCUSSION THREAD

**Discussion by sezeregrek** [tailwindlabs/tailwindcss#2793] [Help] [answered]
Hi all,

I am trying to export tailwind css with purge. In final css file, there isn't any `dark` related classnames. My settings is below. Are there any problems in my settings? Or, is there anyone facing with the same problem?

tailwind.config.js
```	
purge: {
  enabled: true,
  layers: ["components", "utilities"],
  content: ["./src/**/*.tsx"],
},
darkMode: "class"
```

package.json 

```
"postcss": "8.1.1",
"tailwindcss": "2.0.1",
```

```
"build:css": "tailwindcss build tailwind/tailwind.css -o src/common/styles/tailwind.css",
```

**Comment by sezeregrek** [comment]
Any idea?

**Comment by shivamsn97** [comment]
I am having the same problem. It purges all classes used in dark mode.

**Comment by Zh4rsiest** [comment]
I got it working from this post: https://github.com/tailwindlabs/tailwindcss/issues/1675#issuecomment-622934302

The problem most likely is that you have to set **NODE_ENV=production** variable in your **package.json** npm scripts so that it will look something like this:

"scripts": {
    "watch": "npm run dev -- --watch",
    "dev": "webpack --mode development --progress",
    "build": "**NODE_ENV=production** webpack --mode=production --progress"
  },

**Comment by simonswiss** [comment]
@sezeregrek @shivamsn97 is this still an issue for you two?

Any repo where you can reproduce the issue for us to have a look?

**Reply by shivamsn97** [reply]
Yes. This is still an issue for me. Since I have defined purge.enabled = true, so NODE_ENV should not matter. I have described this issue in #2943 and reproduced in https://github.com/shivamsn97/tailwind-bug-reproduction.

**Reply by simonswiss** [reply]
Thanks for that - I'll have a look 👍

**Reply by simonswiss** [reply]
I replied to the issue - but have also posted the answer below 👇 to follow-up on this discussion 👍

**Comment by simonswiss** [comment, marked as answer]
Hey!

I just had a look at @shivamsn97's [repo](https://github.com/shivamsn97/tailwind-bug-reproduction).

When you are using the `darkMode: 'class'` strategy, the generated CSS for the dark mode variants will be:

```css
.dark .dark\:text-gray-200
```

PurgeCSS is looking for the actual `dark` class name in order to keep these classes in the CSS output. If it doesn't (even if it finds `dark:text-gray-200`), it will throw them away.

You can verify that by adding `class="dark"` to the `<body>` tag in your HTML page, and run the build command again. Dark mode classes will now be preserved.

Another approach is to ass `dark` in your safelist for the purge options:

```diff
purge: {
  enabled: true,
  preserveHtmlElements: false,
  content: ["**/*.html", "**/*.js"],
+  options: {
+    safelist: ["dark"],
+  },
},
```

We should probably update the documentation to cover this "gotcha" 👍

**Reply by zbecknell** [reply]
You are a champion!

**Reply by simonswiss** [reply]
By the way, another solution (probably better than safelisting the `dark` classname) would be to add the path to the JS file that toggles the `dark` class in the DOM. 

If the `dark` string is present in that file, and the file is included in the paths of `purge.content`, it should work properly without having to safelist.

**Comment by shivamsn97** [comment]
Thanks. I got this.
